File: config/manager_secret/elixir/lib/serviceradar_secret/secret.ex
defmodule ServiceradarSecret.Secret do
  @moduledoc """
  A resolved secret value, which cannot be inspected.

  **The value is held in a closure, not in a field.** A custom `Inspect` implementation is not
  enough on the BEAM: a struct IS a map, so `inspect(term, structs: false)` renders it as one and
  prints every field, bypassing the protocol. Anything that walks the term rather than calling
  `Inspect` sees the same thing. A closure's environment is not rendered, so the value survives
  `structs: false`, a Logger formatter that flattens metadata, and a crash report that dumps the
  term.

  `Inspect` and `String.Chars` still redact, for the ordinary paths.

  The value is reachable only through `expose/1`. That name is the point: every call site that
  reads the value says so, and a grep for `expose` is the complete list of places a secret can
  leave this struct.
  """

  @redacted "[REDACTED]"
  def redacted, do: @redacted

  @enforce_keys [:reveal]
  defstruct [:reveal]

  @opaque t :: %__MODULE__{reveal: (-> String.t())}

  @doc """
  Wraps a resolved value.

  Empty is not a secret: a provider that returns an empty string has failed to resolve one, and
  treating it as a value is how a component connects with a blank password.
  """
  @spec new(String.t()) :: {:ok, t()} | :error
  def new(""), do: :error
  def new(value) when is_binary(value), do: {:ok, %__MODULE__{reveal: fn -> value end}}

  @doc "The value. Named so that reading it is visible in review and greppable in audit."
  @spec expose(t()) :: String.t()
  def expose(%__MODULE__{reveal: reveal}), do: reveal.()

  @spec length(t()) :: non_neg_integer()
  def length(%__MODULE__{} = secret), do: secret |> expose() |> byte_size()
end

defimpl Inspect, for: ServiceradarSecret.Secret do
  import Inspect.Algebra

  def inspect(_secret, _opts) do
    concat(["#Secret<", ServiceradarSecret.Secret.redacted(), ">"])
  end
end

defimpl String.Chars, for: ServiceradarSecret.Secret do
  def to_string(_secret), do: ServiceradarSecret.Secret.redacted()
end

File: config/manager_secret/go/secret.go
// Package secret resolves a component's secrets through the provider its environment selects.
//
// The environment comes from SERVICERADAR_ENV, the same single variable ConfigManager reads; a
// component never names a provider. Logical secret names are identical across environments and
// languages, so only the provider changes.
//
// Two properties are load-bearing and both are enforced by the type rather than by discipline: a
// Secret cannot be printed, and a name a component did not declare is refused before the provider
// is consulted.
package secret

// Redacted is what a secret renders as in both String and GoString.
const Redacted = "[REDACTED]"

// Secret is a resolved secret value, which cannot be printed.
//
// String and GoString are implemented to redact, and the value is reachable only through Expose.
// That name is the point: every call site that reads the value says so, and a grep for Expose is
// the complete list of places a secret can leave this type.
//
// The field is unexported so %v and %+v on the struct cannot reach it, and GoString covers %#v,
// which would otherwise print the struct literal including the value.
type Secret struct {
	value string
}

// NewSecret wraps a resolved value.
//
// Empty is not a secret: a provider that returns an empty string has failed to resolve one, and
// treating it as a value is how a component connects with a blank password.
func NewSecret(value string) (Secret, bool) {
	if value == "" {
		return Secret{}, false
	}
	return Secret{value: value}, true
}

// Expose returns the value. Named so that reading it is visible in review and greppable in audit.
func (s Secret) Expose() string { return s.value }

func (s Secret) Len() int { return len(s.value) }

// String redacts. fmt reaches this for %v and %s.
func (s Secret) String() string { return Redacted }

// GoString redacts. Without it %#v prints the struct literal, value included -- and %#v is what
// a debugger dump or a struct-printing logger reaches for.
func (s Secret) GoString() string { return "Secret(" + Redacted + ")" }
